cou/apps/auxiliary.py

```diff
@@ -431,6 +431,51 @@ def upgrade_plan_sanity_checks(
 class OVNPrincipal(OVN):
     """OVN principal application class."""
 
+    def pre_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PreUpgradeStep]:
+        """Pre Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        steps = super().pre_upgrade_steps(target, units)
+        if self.config.get("enable-auto-restarts", {}).get("value") is False:
+            steps.extend(self.get_run_deferred_hooks_and_restart_pre_upgrade_step(units))
+        return steps
+
+    def post_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PostUpgradeStep]:
+        """Post Upgrade steps planning.
+
+        Wait until the application reaches the idle state and then check the target workload.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate post upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of post upgrade steps.
+        :rtype: list[PostUpgradeStep]
+        """
+        steps = []
+        if self.config.get("enable-auto-restarts", {}).get("value") is False:
+            steps.extend(self.get_run_deferred_hooks_and_restart_post_upgrade_step(units))
+        steps.extend(super().post_upgrade_steps(target, units))
+        return steps
+
+    def _check_auto_restarts(self) -> None:
```

Comment:
**nitpick:** 

Instead of override the method, can the behavior of function change base on the class attribute?

```
class OpenstackApplication:

    check_auto_restarts: bool = True
    
    def _check_auto_restarts(self) -> None:
        if self.check_auto_restarts:
            ...
            
```

Reply:
I think this check will be gone when we support all charms that has `enable-auto-restarts` (ovn-dedicated-chassis, ovn-central, ovn-chassis, rabbitmq-server), so it's not important to change the behavior now